packages/backend/src/nest/socket/socket.service.ts

```diff
@@ -195,25 +201,72 @@ export class SocketService extends EventEmitter implements OnModuleInit {
         this.emit(SocketActionTypes.LOAD_MIGRATION_DATA, data)
       })
     })
+
+    // Ensure the underlying connections get closed. See:
+    // https://github.com/socketio/socket.io/issues/1602
+    //
+    // I also tried `this.serverIoProvider.io.disconnectSockets(true)`
+    // which didn't work for me.
+    const sockets = new Set<net.Socket>()
+
+    this.serverIoProvider.server.on('connection', conn => {
+      sockets.add(conn)
+      conn.on('close', () => {
+        sockets.delete(conn)
+      })
+    })
+
+    return () => sockets.forEach(s => s.destroy())
+  }
+
+  public getConnections = (): Promise<number> => {
+    return new Promise(resolve => {
+      this.serverIoProvider.server.getConnections((err, count) => {
+        if (err) throw new Error(err.message)
+        resolve(count)
+      })
+    })
   }
 
   public listen = async (port = this.configOptions.socketIOPort): Promise<void> => {
-    return await new Promise(resolve => {
-      if (this.serverIoProvider.server.listening) resolve()
+    this.logger(`Opening data server on port ${this.configOptions.socketIOPort}`)
+
+    // Sometimes socket.io closes the HTTP server but doesn't close
+    // all underlying connections. So it doesn't appear that
+    // `this.serverIoProvider.server.listening` is sufficient.
+    if (this.listening) {
+      const numConnections = await this.getConnections()
+      this.logger('Failed to listen. Connections still open:', numConnections)
+      return
+    }
+
+    return new Promise(resolve => {
       this.serverIoProvider.server.listen(this.configOptions.socketIOPort, '127.0.0.1', () => {
         this.logger(`Data server running on port ${this.configOptions.socketIOPort}`)
+        this.listening = true
         resolve()
       })
     })
   }
 
-  public close = async (): Promise<void> => {
-    this.logger(`Closing data server on port ${this.configOptions.socketIOPort}`)
-    return await new Promise(resolve => {
-      this.serverIoProvider.server.close(err => {
+  public close = (): Promise<void> => {
+    return new Promise(resolve => {
+      this.logger(`Closing data server on port ${this.configOptions.socketIOPort}`)
+
+      if (!this.listening) {
+        this.logger('Data server is not running.')
+        resolve()
+        return
+      }
+
+      this.serverIoProvider.io.close(err => {
         if (err) throw new Error(err.message)
+        this.logger('Data server closed')
+        this.listening = false
         resolve()
       })
+      this.logger('Disconnecting sockets')
+      this.closeSockets()
```

Comment:
[question] Shouldn't we also detach listeners when closing sockets? I mentioned it before, but I'd rather see a dedicated `closeSockets()` and `detachListeners()` method.

Reply:
I don't think we need to detach any listeners.